File: src/main/java/com/dude/dms/ui/components/tags/AttributeValueContainer.java
package com.dude.dms.ui.components.tags;

import com.dude.dms.backend.brain.DmsLogger;
import com.dude.dms.backend.data.docs.AttributeValue;
import com.dude.dms.backend.data.docs.Doc;
import com.dude.dms.ui.builder.BuilderFactory;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;

import java.util.ArrayList;
import java.util.List;

public class AttributeValueContainer extends VerticalLayout {

    private static final DmsLogger LOGGER = DmsLogger.getLogger(AttributeValueContainer.class);

    private final List<AttributeValueField> fields;

    public AttributeValueContainer(BuilderFactory builderFactory, Doc doc) {
        fields = new ArrayList<>();

        setPadding(false);
        setSpacing(false);
        for (AttributeValue attributeValue : doc.getAttributeValues()) {
            AttributeValueField field = builderFactory.attributes().valueField(attributeValue).build();
            field.setWidthFull();
            add(field);
            fields.add(field);
        }
    }

    public boolean validate() {
        for (AttributeValueField field : fields) {
            if (!field.validate()) {
                LOGGER.showError("Attribute '" + field.getLabel() + " is required.");
                return false;
            }
        }
        return true;
    }
}

File: src/main/java/com/dude/dms/ui/components/tags/AttributeValueContainer.kt
package com.dude.dms.ui.components.tags

import com.dude.dms.backend.containers.DocContainer
import com.dude.dms.brain.DmsLogger
import com.dude.dms.ui.builder.BuilderFactory
import com.vaadin.flow.component.UI
import com.vaadin.flow.component.html.Div

class AttributeValueContainer(private val builderFactory: BuilderFactory, private val readOnly: Boolean = false) : Div() {

    private val fields = mutableListOf<AttributeValueField>()

    var onChange: (() -> Unit)? = null

    init {
        addClassName("attribute-container")
    }

    fun fill(docContainer: DocContainer) {
        clear()
        docContainer.attributeValues.forEach {
            val field = builderFactory.attributes().valueField(it, readOnly).also { field ->
                field.setWidthFull()
                field.onChange = onChange
            }
            add(field)
            fields.add(field)
        }
    }

    fun clear() {
        removeAll()
        fields.clear()
    }

    fun validate(silent: Boolean = false): Boolean {
        for (field in fields) {
            if (!field.validate()) {
                if (!silent) LOGGER.showError("Attribute '" + field.label + " is required.", UI.getCurrent())
                return false
            }
        }
        return true
    }

    companion object {
        private val LOGGER = DmsLogger.getLogger(AttributeValueContainer::class.java)
    }
}